### PR TITLE
Improve UX when adding/editing records

### LINKED ISSUE #549: Improve UX for adding records
Right now, when a user is adding or modifying a record, is the user hits enter or clicks next / previous, data will be lost. 

Scenario 1
Current Behavior:
User clicks add / edit
User fill out information
User presses enter
Data is lost

Expected Behavior:
User clicks add / edit
User fill out information
User presses enter
Record is saved

Scenario 2
Current Behavior:
User clicks add / edit
User fill out information
User clicks prev or next
Data is lost

Expected Behavior:
User clicks add / edit
User fill out information
User clicks prev or next
User is prompted that data will be lost if record is in dirty state.

### PR DESCRIPTION
**Pull request by dbelokon**
**What kind of change does this PR introduce?**
- Other, please describe: UX improvement
- Minor test adjustments

Fixes #549.

I also had to modify certain tests because some of them would fail due to the usage of `<Prompt>` outside of `<Router>` component. I went ahead and modified those tests so that the test subjects are wrapped by a `<BrowserRouter>`. The error shown was very similar to [this error discussed in a StackOverflow discussion](https://stackoverflow.com/questions/62710458/invariant-failed-you-should-not-use-switch-outside-a-router).

### CODE REVIEW

File: src/components/FormContainer.tsx [hunk 1]
```diff
@@ -263,11 +296,19 @@ const FormListContainer = <A,>(
 
   return (
     <>
+      <Prompt
+        when={formState === FormState.Dirty}
```
Comment (zakpatterson) [inline src/components/FormContainer.tsx:300]: Did you run across `react-hook-form`'s `isDirty`? 

It should be in scope already, in the above line:

`const { reset, handleSubmit, formState: {isDirty} } = useFormContext()`

Maybe you already saw that and there's a reason it doesn't fit this situation?

**Comment by thegrims** [comment]
Created a new issue [If foreign address is checked, it incorrectly remains checked even if it is unchecked #700](https://github.com/ustaxes/UsTaxes/issues/700) in GitHub

File: src/components/FormContainer.tsx [hunk 2]
```diff
@@ -263,11 +296,19 @@ const FormListContainer = <A,>(
...
+        message="Warning: you have unsaved changes. Are you sure you want to leave?"
```
Comment (thegrims) [inline src/components/FormContainer.tsx:301]: I'm not seeing this message show up in the preview environment, is there a section/page I should visit?

Comment (zakpatterson) [reply src/components/FormContainer.tsx:301]: It looks like it just shows up on pages with the internal item forms, like 1099s, W-2s, spouseanddependent, if the user navigates away using the sidebar while unsaved data is present in the form.

image

Comment (thegrims) [reply src/components/FormContainer.tsx:301]: Gotcha

File: src/components/FormContainer.tsx (hunk 1 above)
Comment (zakpatterson) [reply src/components/FormContainer.tsx]: I went ahead and switched the implementation to use the existing `FormContext`'s `isDirty`, and it seems to work the same.

**Review by zakpatterson — APPROVED** [review]

**Review by thegrims — APPROVED** [review]
LGTM

**Comment by thegrims** [comment]
@dbelokon Thanks for your contribution!

**Comment by dbelokon** [comment]
You are welcome!! 😁

**Pull request merged**